lib/developer_portal/app/controllers/developer_portal/login_controller.rb

```diff
@@ -25,8 +26,9 @@ def new
 
   def create
     logout_keeping_session!
+    return render_creation_error('Spam protection failed.') unless spam_check(buyer)
```

Comment:
In my opinion, we must do a spam check before we are authenticating the user (so they can't DDoS the system).

The problem here is what model do we pass to the `spam_check`?

- `nil` - does make sense, but it's weird in my opinion
- dummy model - similar to `nil`, but it makes more sense in my opinion. However, the object is marked as spam with a 1.0 probability. That's why I added the check if the session is marked as spam.
- we find the user based on his username/email - we can't do this because it can be used for the DDoS.

In conclusion, I chose to go with the dummy model.

Reply:
> Does this mean user will be asked for captcha before every single login?
Technically no, because we have 3 settings for the captcha (None, Suspicious only, and Always).
Suspicious only should decide based on our own checks if we trigger reCaptcha or not.

> Or that we just leave reCaptcha decide whether to ask user for something or not based on its own algorithms?
We are only checking if it passed the reCaptcha.

Reply:
If there is an active attempt to brute-force a password or automate login, then the process will probably create a clean session by every attempt to login. So I'm not sure if this would effectively eliminate the captcha for bots and enable it only for real users mistyping their passwords.

I think such logic needs to be coordinated with somebody with experience with recaptcha.